pkg/common_controller/snapshot_controller.go

```diff
@@ -0,0 +1,1237 @@
+	ctrl.checkandAddSnapshotFinalizers(snapshot, snapshotBound, deleteContent)
+
+	klog.V(5).Infof("processFinalizersandDeleteContent[%s]: check if we should remove finalizer on snapshot source and remove it if we can", utils.SnapshotKey(snapshot))
+
+	// Check if we should remove finalizer on PVC and remove it if we can.
+	errFinalizer := ctrl.checkandRemovePVCFinalizer(snapshot)
+	if errFinalizer != nil {
+		klog.Errorf("error check and remove PVC finalizer for snapshot [%s]: %v", snapshot.Name, errFinalizer)
+		// Log an event and keep the original error from syncUnready/ReadySnapshot
+		ctrl.eventRecorder.Event(snapshot, v1.EventTypeWarning, "ErrorPVCFinalizer", "Error check and remove PVC Finalizer for VolumeSnapshot")
+	}
+	return nil
+}
+
+// deleteContentandSnapshotFinalizers deletes the content and removes snapshot finalizers if needed
+func (ctrl *csiSnapshotCommonController) deleteContentandSnapshotFinalizers(snapshot *crdv1.VolumeSnapshot, content *crdv1.VolumeSnapshotContent, deleteContent bool) error {
+	klog.V(5).Infof("deleteContentAndSnapshotFinalizers VolumeSnapshot[%s]: %s", utils.SnapshotKey(snapshot), utils.GetSnapshotStatusForLogging(snapshot))
+
+	var err error
+	// Check is snapshot deletionTimestamp is set and any finalizer is on
+	if utils.IsSnapshotDeletionCandidate(snapshot) {
+		// Volume snapshot should be deleted. Check if it's used
+		// and remove finalizer if it's not.
+		// Check if a volume is being created from snapshot.
+		inUse := ctrl.isVolumeBeingCreatedFromSnapshot(snapshot)
+
+		klog.V(5).Infof("syncSnapshot[%s]: set DeletionTimeStamp on content.", utils.SnapshotKey(snapshot))
+		// If content exists, set DeletionTimeStamp on the content;
+		// content won't be deleted immediately due to the finalizer
+		if content != nil && deleteContent {
+			err := ctrl.clientset.SnapshotV1beta1().VolumeSnapshotContents().Delete(content.Name, &metav1.DeleteOptions{})
+			if err != nil {
+				ctrl.eventRecorder.Event(snapshot, v1.EventTypeWarning, "SnapshotContentObjectDeleteError", "Failed to delete snapshot content API object")
+				return fmt.Errorf("failed to delete VolumeSnapshotContent %s from API server: %q", content.Name, err)
+
+			}
+		}
+
+		if !inUse || (content == nil && err == nil) {
+			klog.V(5).Infof("syncSnapshot: Remove Finalizer for VolumeSnapshot[%s]", utils.SnapshotKey(snapshot))
+			doesContentExist := false
+			if content != nil {
+				doesContentExist = true
+			}
+			return ctrl.removeSnapshotFinalizer(snapshot, !inUse, !doesContentExist)
+		}
+	}
+	return nil
+}
+
+// checkandAddSnapshotFinalizers checks and adds snapshot finailzers when needed
+func (ctrl *csiSnapshotCommonController) checkandAddSnapshotFinalizers(snapshot *crdv1.VolumeSnapshot, snapshotBound bool, deleteContent bool) {
+	addSourceFinalizer := false
+	addBoundFinalizer := false
+	if utils.NeedToAddSnapshotAsSourceFinalizer(snapshot) {
```

Comment:
Should not this finalizer being added only when a VolumeSnapshot object is being used as source to restore a volume?
It seems to me it will be always added in this routine regardless whether it's under using or not.